### PR TITLE
[SRK-55] Actualize workflow.md

### PR DESCRIPTION
**Pull request by gromakovsky**
I tried to rewrite some parts of `workflow.md` to make it correspond to our actual workflow. Some changes are subjective and comments are welcome.

### CODE REVIEW

**Review by chshersh — APPROVED** [review]

**Comment by jagajaga** [comment]
@gromakovsky check new pic

File: workflow.md
Comment (jagajaga) [inline workflow.md]: Not true. explorer can have any version it wants (it's another product).
@akegalj @sectore please help @gromakovsky to write this section.

File: workflow.md
Comment (jagajaga) [inline workflow.md]: No, `mid` branch can be with any release branch

**Review by jagajaga — CHANGES_REQUESTED** [review]

**Comment by neongreen** [comment]
@jagajaga this label got lost: 

image

**Comment by jagajaga** [comment]
@neongreen no, because we have only `X.Й` branches now

File: workflow.md
Comment (gromakovsky) [reply workflow.md]: Then I need an example. Suppose there is `cardano-sl-0.8` already, what's the point of having `cardano-sl-0.5-mid` in this case, for example?

I see how `cardano-sl-0.8-mid` is useful, it can be used by latest explorer (`master`). It is basically a stable version (`cardano-sl-0.8`) + some changes from explorer guys who know what they are doing.

But I don't understand why `cardano-sl-0.5-mid` can be useful. Where is it suppose to be used? 
By explorer's release? Suppose there is explorer's release 0.2 which uses cardano-sl release 0.5. Explorer should point to some commit in `cardano-sl-0.5`. If we want to change something in explorer 0.2 which requires changes in `cardano-sl-0.5`, we should make a PR to `cardano-sl-0.5` (not `-mid`), right? It's still obscure to me whether `-mid` branch is only for master or for release branches.
By explorer's master? If explorer lags behind master so much, then we probably want it to catch up ASAP, don't we? On the other hand, I can imagine situation that `cardano-sl-0.8` has substantial changes in API comparing to `cardano-sl-0.5` and we don't want to spend time on moving explorer to `cardano-sl-0.5`, so master is still using `-0.5`. That's the only case I see where `-0.5-mid` can be useful. Is it indeed the only case? If it is, I will write about it in `workflow.md`, because it's not obvious and it will make it easier to understand this workflow.

File: workflow.md
Comment (ksaric) [inline workflow.md]: So do we use the `-mid` branch for _both_ `cardano-sl-explorer` and `daedalus`?

File: workflow.md
Comment (ksaric) [reply workflow.md]: If it's a case that we don't need multiple `-mid` branches (which as @gromakovsky says, doesn't have much sense) could we just have a single branch called `cardano-sl-mid` and tag the releases? If what we have is _only_ the latest version, there shouldn't be any problem with fixes.

File: workflow.md
Comment (sectore) [inline workflow.md]: What is a "big", what is a "small" breaking change? And why do we add breaking changes into a `Min` version? I would prefer to have API breaking in `Maj` only (as before). With this `Maj` is not just useful for marketing, but also for all the other users which are creating other applications or tools depending on an API.

File: workflow.md
Comment (sectore) [inline workflow.md]: I think it is hard to say that `feature A` is a minor change (`Fix`) and `feature B` is not (`Min`). That might be subjective. What about to mark any feature (minor, small, large) w/o breaking changes as `Min`? Just to simplify any personal decision.

**Review by sectore — CHANGES_REQUESTED** [review]
@gromakovsky Fantastic documentation! I have added just few questions + one suggestion.

File: workflow.md
Comment (gromakovsky) [reply workflow.md]: Yes, that was my initial understanding and I still think so. Maybe @georgeee had a different vision when he proposed it.

File: workflow.md
Comment (gromakovsky) [reply workflow.md]: > If it's a case that we don't need multiple `-mid` branches

@jagajaga says that it's not the case, so probably we can't.

As I wrote in my big comment (which probably nobody will read because it's too big 😜), I see only one case when it may be useful to have multiple `-mid` branches: when some project (e. g. explorer) doesn't support the latest version of `cardano-sl`. However, even in this case I don't see why it may be necessary to have `-mid` branch associated with some old version. If explorer team wants to change something in the old version, they can just make a PR to branch with old version, not `-mid` branch associated with this old version.

So I would like to have some clarification about why we need to have multiple `-mid` branches.

File: workflow.md
Comment (gromakovsky) [reply workflow.md]: There are different APIs: API exposed by haskell code (modules, functions), API used for communication between nodes (TCP), web API, maybe something else. And there are different changes by their significance. For example, if someone renames one function in haskell code it is much less significant change than if someone breaks API for node-node communication (imo). The former change is visible only to small group of developers while the second one may be visible to much broader group of people (e. g. developers of alternative implementations or some custom clients). So I think it would be good if such changes corresponded to different parts of version.

Maybe we should use 4 parts instead of 3 if we don't manage to describe versioning policy using only 3 parts.

File: workflow.md
Comment (gromakovsky) [reply workflow.md]: Here I initially wrote `when something minor is added, insignificant from end-user perspective` imagining cases when we have a release and we want to:
• add some logging because we know there is a bug somewhere and we have difficulties investigating it
or
• export some function from some module because it suddenly turned to be needed by explorer or whatever else
or
• improve something in documentation (e. g. haddock) because people read it and we want it to be good and some people complained they don't understand something (well, maybe it can be treated as a fix already)
etc. These are not really fixes (especially adding more logging), but on the other hand these changes seem to small to increase `Min` version (because in CSL case it is a long process).

Again, maybe we should switch to versioning with 4 parts to resolve it.

General rule is: if it's an important change which is necessary in X.Y, then just add it to X.Y. If it's not, then don't add it to X.Y.

File: workflow.md
Comment (sectore) [reply workflow.md]: @jagajaga @gromakovsky 

Here is my proposal to update chapter `Interaction between cardano-sl and projects dependent on it`

```
### Interaction between cardano-sl and projects dependent on it

Some projects (e. g. `cardano-sl-explorer`, `daedalus`) depend on
`cardano-sl`. In order to facilitate interaction with `cardano-sl` we
use special branch `-mid` mentioned above. `Explorer` is used as an
example in the rest of this subsection, but rules apply to everything
dependent on `cardano-sl`.

* `Explorer` must depend on version of `cardano-sl` from `-mid` branch only,
  not `-rc` or release branch. This rule is imposed
  because only this branch is generally stable and
  well-tested for using it by third party applications.
  `-rc` branch corresponds to a version which is not
  properly tested yet, hence explorer should not depend on it.
* `Explorer` comes with its own version based on [CSL versioning](Versions).
  However, to reflect current version of `cardano-sl`, Explorer is using a branch
  `cardano-sl-X.Y`, where `X.Y` is `Maj.Min` of `cardano-sl-X.Y-mid`.
  That's because different versions of `cardano-sl` may be incompatible in some way and it may
  lead to different (inconsistent) behavior of `cardano-node` and
  `cardano-explorer`. For example: `Explorer's` branch `cardano-sl-0.4` is pointing to `0.4` of `cardano-sl`.

* If `Explorer` developer wants to change something in `cardano-sl` for
  `Explorer's` version `X.Y.Z` they should make a PR to `cardano-sl-X.Y`
  branch, wait until it's merged and then use newer version in
  explorer.
* `master` branch of `Explorer` should depend on the latest `-mid`
  branch of `cardano-sl`.
* If `Explorer` developers wants to change something in `cardano-sl` for
  explorer's `master` they should make a PR to `cardano-sl-X.Y-mid` of
  `cardano-sl` (where `X.Y` is latest version of `cardano-sl`).
* `cardano-sl-X.Y.` should be frequently merged into
  `cardano-sl-X.Y-mid`.
* `cardano-sl-X.Y-mid` should be frequently merged to `master`.
```

Comment (jagajaga) [reply workflow.md]: @sectore I disagree with ``Explorer` must depend on version of `cardano-sl` from `-mid` branch only, not `-rc` or release branch.` It can depend on `mid` or release branch.

File: workflow.md
Comment (ksaric) [reply workflow.md]: > when some project (e. g. explorer) doesn't support the latest version of cardano-sl

Isn't that then the time *between the old version and the new version*? Why do we need to track this? Isn't this something we need to agree on before pushing onto that branch anyway since it seems it's going to be for both `explorer` and `daedalus`? 
Like - "in the new `-mid` version there will be HD wallets and multiple payments"?

>  If explorer team wants to change something in the old version

Am I missing something - don't we plan to have a single released public version of `explorer`?

Comment (gromakovsky) [reply workflow.md]: About the first part: I am not sure I understand what you mean, but I can clarify what I mean. I was imagining a situation when, for example, there is version 0.7 of cardano-sl supported by explorer and then there is a version 0.8 which doesn't break API used for communication between explorer and CSL, but it contains a massive refactoring which changes module structure, various types, etc. And at the same time there are some pressing issues on explorer. In this case we may want not postpone using 0.8 in explorer and finish with the most pressing issues first. In this case it may make sense to have `0.7-mid` even though 0.8 exists.
But I think I see your point, usually we will have to support latest version of cardano-sl in explorer even if old version is compatible. But maybe sometimes it won't be the case, for example, if we add some feature to core and it won't be a priority to display it in explorer.

> Am I missing something - don't we plan to have a single released public version of explorer?

I think we do, but there I meant situation when explorer team wants to change something in old version of **cardano-sl**. As in example above, suppose there is CSL-0.8, but explorer uses 0.7 (which should be generally avoided, but suppose it was necessary). And explorer team may want to change something in CSL-0.7. I think it should be changed in `cardano-sl-0.7`, not in `cardano-sl-0.7-mid`.

File: workflow.md
Comment (gromakovsky) [reply workflow.md]: > However, to reflect current version of `cardano-sl`, Explorer is using a branch
  `cardano-sl-X.Y`, where `X.Y` is `Maj.Min` of `cardano-sl-X.Y-mid`.

So `cardano-sl-X.Y` branch in Explorer is basically release branch, right? It implies that each `X.Y` version of explorer depends on `X.Y` of `cardano-sl`, doesn't it? That's what I initially proposed, but @jagajaga disagrees.

> If `Explorer` developer wants to change something in `cardano-sl` for
  `Explorer's` version `X.Y.Z` they should make a PR to `cardano-sl-X.Y`
  branch, wait until it's merged and then use newer version in
  explorer.

So PRs are made to `-X.Y` branch, but Explorer used `-X.Y-mid` branch. What's the purpose of `-mid` branch then?

File: workflow.md
Comment (gromakovsky) [reply workflow.md]: > Am I missing something - don't we plan to have a single released public version of explorer?

Also, @ksaric, please don't forget that this section is not only about Explorer, but actually about everything that depends on CSL. I wrote about it in the file:
> Explorer is used as an example in the rest of this subsection, but rules apply to everything dependent on `cardano-sl`.

And I think it would be good to document long-term workflow and nobody knows how many projects we will have in future. Currently we also have `daedalus` apart from `explorer`. And for `deadalus` we whould like to support all versions which people use.

**Comment by jagajaga** [comment]
ping all

File: workflow.md
Comment (Martoon-00) [inline workflow.md:65]: What is "AST of the program"?

Comment (gromakovsky) [reply workflow.md:65]: Abstract syntax tree. For example, when you add a comment or split one line into multiple lines, etc.

I don't know how to formulate it better. I think that we also don't need a special issue if we just fix a typo in a variable name which is used only in 1 or 2 places, for example. But I don't know how to describe it formally.

Comment (Martoon-00) [reply workflow.md:65]: I see, thanks.
It's on your responsibility then to answer everybody who asks about this ))

**Pull request merged**
